File: dynamic-programming/322-coin-change/20260806/AISolution/322.coin-change.java
import java.util.Arrays;

// 問題: 322. Coin Change
// アプローチ: 金額ごとの最小硬貨枚数を記録するボトムアップ動的計画法
// 時間計算量: O(amount * coins.length)
// 空間計算量: O(amount)
class Solution {
    public int coinChange(int[] coins, int amount) {
        // amount + 1 は、作れる金額で必要になる最大枚数より大きい値である。
        int unreachable = amount + 1;
        int[] minimumCoins = new int[amount + 1];
        Arrays.fill(minimumCoins, unreachable);
        minimumCoins[0] = 0;

        // 各金額について、最後に使う硬貨をすべて試す。
        for (int currentAmount = 1; currentAmount <= amount; currentAmount++) {
            for (int coin : coins) {
                if (coin <= currentAmount) {
                    minimumCoins[currentAmount] = Math.min(
                        minimumCoins[currentAmount],
                        minimumCoins[currentAmount - coin] + 1
                    );
                }
            }
        }

        return minimumCoins[amount] == unreachable ? -1 : minimumCoins[amount];
    }
}

File: dynamic-programming/322-coin-change/20260806/FirstSolution/322.coin-change.cs
public class Solution
{
    // 最小枚数を返すとあるので、大きい額から使えないか探すと良さそう？
    // 目標額amountが0になるケースはそのまま0で返せば良さそう
    // 額面を作れない場合は -1 を返す必要がある
    // 無限に額面を使える
    // 制約
    // - 1 <= coins.length <= 12
    // - 1 <= coins[i] <= 2^31 - 1
    // - 0 <= amount <= 10^4
    public int CoinChange(int[] coins, int amount)
    {
        int unreachable = amount + 1;

        // 各金額ごとにその金額を構成する最小枚数を調べる
        int[] minCoins = new int[amount + 1]; // 0~amount yenまでの金額でそれぞれ最小で必要な枚数を記録
        Array.Fill(minCoins, unreachable);
        minCoins[0] = 0;

        for (int currentAmount = 1; currentAmount <= amount; currentAmount++)
        {
            // コインの種別ごとに、目標額に達する最後がそのコインだった場合、何枚必要かを計算していく
            foreach (var coin in coins)
            {
                if (coin <= currentAmount)
                {
                    minCoins[currentAmount] = Math.Min(minCoins[currentAmount], minCoins[currentAmount - coin] + 1);
                }
            }
        }

        return minCoins[amount] == unreachable ? -1 : minCoins[amount];
    }
}
